igraph/igraph
# Implement umap

I've started trying to implement UMAP. There's so much new stuff, that I'm writing code just to explore the problem and see what we need. 

**Prototype iteration summary:**
- Initially start from data in a vector space: perhaps skip graph building and start directly with weighted graph (e.g. knn) in a later iteration
- Start directly with a sparse graph with "distances"
- Implement rough functions
- Implement conversion distance->probability in original graph, `mu(a)`
- Implement smoothed Euclidean distance->probability in embedded space, `nu(a)`, via Gauss-Newton fitting of hyperparameters a and b (using line search to deal with strong off-diagonal elements of the Hessian)
- Implement binary search for `sigma`, a rough estimate of the decay of the neighbor distance beyond the closest neighbor
- Implement corner cases for the `sigma` calculation
- Implement cross-entropy (CE) function
- Implement derivatives of the CE for attractive and repulsive forces
- Double check everything
- Put it all together
- Write header file
- Write unit test code
- Write unit test output
- Test produces reasonable results, no strict checking yet ('t's a stochastic algo, so not sure how to do that?)
- Check function naming
- Write function docstring (first draft)
- Tests pass locally (Fabio's computer)
- Add a bunch of brackets that depend on `#define UMAP_DEBUG` for printing stuff... is this any good?
- Redo the math on the CE and derivatives to make sure we are doing it correctly
 - Add unit test for unweighted graph
 - Lint code and docs
 - 3D (theoretically ND) version
 - Cover trivial graphs
 - improve docs on `(a,b)` fit scaling behaviour
 - Stress test under a few corner and not-so-corner cases
- Ask other igraph folks for review/suggestions
- Move to `layout`

Todo:
- Decide what to do about initializatoin choices
- Merge PR :smirk:

Possible optimization:
- Use openMP

edit (Fabio): added brief idea list as requested by @szhorvat
